python/mxnet/ndarray/ndarray.py

```diff
@@ -1703,6 +1912,109 @@ def tostype(self, stype):
+    elif isinstance(key, list):
+        # TODO(junwu): Add support for nested lists besides integer list
+        for i in key:
+            if not isinstance(i, integer_types):
+                raise TypeError('Indexing NDArray only supports a list of integers as index'
+                                ' when key is of list type, received element=%s of type=%s'
+                                % (str(i), str(type(i))))
+        return _ADVANCED_INDEXING
+    elif isinstance(key, integer_types):
+        return _BASIC_INDEXING_STEP_EQUAL_ONE
+    elif isinstance(key, tuple):
+        step_not_one = False
+        for idx in key:
+            if isinstance(idx, (NDArray, np.ndarray, list, tuple)):
+                return _ADVANCED_INDEXING
+            elif isinstance(idx, py_slice):
+                if idx.step is not None and idx.step != 1:
+                    step_not_one = True
+            elif not isinstance(idx, integer_types):
+                raise ValueError("NDArray does not support slicing with key %s of type %s."
+                                 % (str(idx), str(type(idx))))
+        if step_not_one:
+            return _BASIC_INDEXING_STEP_NOT_ONE
+        return _BASIC_INDEXING_STEP_EQUAL_ONE
+    else:
+        return _NDARRAY_INDEXING_UNSUPPORTED
+
+
+def _get_index_range(start, stop, length, step=1):
+    """Given start, stop, step and array length, return
+    absolute values of start, stop, and step for generating index range.
+    The returned values have been compensated by adding length if they
+    are less than zero for all the cases but slice(None, None, -1).
+    Note that the returned value of stop is not necessarily >= 0, since
+    absolute stop is -1 in the case of slice(None, None, -1)."""
+    if step == 0:
+        raise ValueError('step size cannot be zero')
+    if length < 0:
+        raise ValueError('array length cannot be less than zero')
+    if step is None:
+        step = 1
+    if start is None:
+        if step > 0:
+            start = 0
+        else:
+            start = length - 1
+    elif start < 0:
+        start += length
+        if start < 0:
+            raise IndexError('Slicing start %d exceeds limit of %d' % (start-length, length))
+    elif start >= length:
+        raise IndexError('Slicing start %d exceeds limit of %d' % (start, length))
+
+    if stop is None:
+        if step > 0:
+            stop = length
+        else:
+            # this supports case such as ::-1
+            # stop = -1 here refers to the element before index 0,
+            # instead of the last element in the array
+            stop = -1
+    elif stop < 0:
+        stop += length
+        if stop < 0:
+            raise IndexError('Slicing stop %d exceeds limit of %d' % (stop-length, length))
+    elif stop > length:
+        raise IndexError('Slicing stop %d exceeds limit of %d' % (stop, length))
+
+    return start, stop, step
+
+
+def _get_broadcast_shape(shape1, shape2):
```

Comment:
Does numpy have a util for this?

Reply:
I hope so, but I tried searching online and couldn't find one in numpy taking shapes as input arguments.

Reply:
If that is of any use, there is a private function in stride_tricks:
https://github.com/numpy/numpy/blob/77f9540f277de3e727d696eb5987a0505ed8cdf9/numpy/lib/stride_tricks.py#L176

Reply:
@jmargeta Thanks for the information. It looks like that the function in Numpy literally broadcasts ndarrays to get the final shape, which is not efficient and a overkill in this case. Also when the current MXNet NDArray is allocated on GPU, that function does not work. What we need is only the final shape as an input to the MXNet broadcast_to operator which is more efficient and works on both CPU and GPU.